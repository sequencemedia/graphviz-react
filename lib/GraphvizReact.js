import React, { useState, useEffect, useCallback } from 'react';
import classnames from 'classnames';
import { graphviz } from 'd3-graphviz';
import debug from 'debug';
const DEFAULT_OPTIONS = {
    useWorker: false
};
function handleEntries(entries) {
    for (const entry of entries) {
        if (hasEntryTarget(entry)) {
            const target = getEntryTarget(entry);
            if (target instanceof Element) {
                const svg = target.querySelector('svg');
                if (svg instanceof SVGElement) {
                    const { contentRect: { width, height } } = entry;
                    svg.setAttribute('width', width + 'px');
                    svg.setAttribute('height', height + 'px');
                }
            }
        }
    }
}
function DEFAULT_HANDLE_EVENT() {
    //
}
export function hasEventTarget({ target }) {
    if (target instanceof Element)
        return true;
    return false;
}
export function getEventTarget({ target }) {
    if (target instanceof Element)
        return target;
    return null;
}
export function hasEntryTarget({ target }) {
    if (target instanceof Element)
        return true;
    return false;
}
export function getEntryTarget({ target }) {
    if (target instanceof Element)
        return target;
    return null;
}
export function hasCurrent(ref = { current: null }) {
    const { current // = null
     } = ref;
    return (current instanceof Element);
}
export function getCurrent({ current }) {
    return current;
}
const resizeObserver = new ResizeObserver(handleEntries);
const log = debug('@sequencemedia/graphviz-react');
export default function GraphvizReact({ graphRef: ref, dot, className, options = DEFAULT_OPTIONS, onStart = DEFAULT_HANDLE_EVENT, onRenderStart = DEFAULT_HANDLE_EVENT, onRenderEnd = DEFAULT_HANDLE_EVENT, onRenderDot = DEFAULT_HANDLE_EVENT, onEnd = DEFAULT_HANDLE_EVENT, onClick = DEFAULT_HANDLE_EVENT }) {
    log('GraphvizReact');
    const [eventEmitter, setEventEmitter] = useState(null);
    useEffect(() => {
        const { fit = false } = options;
        if (fit) {
            if (hasCurrent(ref)) {
                const current = getCurrent(ref);
                resizeObserver.observe(current);
                return () => {
                    resizeObserver.unobserve(current);
                };
            }
        }
    });
    useEffect(() => {
        if (hasCurrent(ref)) {
            const current = getCurrent(ref);
            const OPTIONS = {
                ...DEFAULT_OPTIONS,
                ...options
            };
            const eventEmitter = (graphviz(current, OPTIONS)
                .renderDot(dot, onRenderDot));
            setEventEmitter(eventEmitter);
        }
    }, [
        dot,
        options
    ]);
    useEffect(() => {
        if (eventEmitter !== null) {
            eventEmitter
                .on('start', onStart)
                .on('renderStart', onRenderStart)
                .on('renderEnd', onRenderEnd)
                .on('end', onEnd);
        }
    }, [
        eventEmitter,
        onStart,
        onRenderStart,
        onRenderEnd,
        onEnd
    ]);
    const handleClick = useCallback(function handleClick(event) {
        if (hasEventTarget(event)) {
            const target = getEventTarget(event);
            if (target instanceof Element) {
                if (hasCurrent(ref)) {
                    const current = getCurrent(ref) ?? { contains() { return false; } };
                    if (current.contains(target))
                        onClick(event);
                }
            }
        }
    }, [
        dot,
        options,
        onClick
    ]);
    log(ref);
    return (React.createElement("div", { className: classnames('graphviz', className), onClick: handleClick, ref: ref }));
}
