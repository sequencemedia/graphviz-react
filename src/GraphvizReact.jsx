import React, {
  createRef,
  useState,
  useEffect,
  useCallback
} from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import {
  graphviz
} from 'd3-graphviz'
import debug from 'debug'

const DEFAULT_OPTIONS = {
  useWorker: false
}

function handleEntries (entries) {
  for (const entry of entries) {
    if (hasEntryTarget(entry)) {
      const target = getEntryTarget(entry)

      const svg = target.querySelector('svg')

      if (svg instanceof SVGElement) {
        const {
          contentRect: {
            width,
            height
          }
        } = entry

        svg.setAttribute('width', width + 'px')
        svg.setAttribute('height', height + 'px')
      }
    }
  }
}

function DEFAULT_HANDLE_START () {
  //
}

function DEFAULT_HANDLE_RENDER_START () {
  //
}

function DEFAULT_HANDLE_RENDER_END () {
  //
}

function DEFAULT_HANDLE_END () {
  //
}

function DEFAULT_HANDLE_CLICK () {
  //
}

export function hasEventTarget ({ target = null }) {
  return (target instanceof Element)
}

export function getEventTarget ({ target = null }) {
  if (target instanceof Element) return target
  throw new Error('Target is not an Element')
}

export function hasEntryTarget ({ target = null }) {
  return (target instanceof Element)
}

export function getEntryTarget ({ target = null }) {
  if (target instanceof Element) return target
  throw new Error('Target is not an Element')
}

export function hasCurrent ({ current = null }) {
  return (current instanceof Element)
}

export function getCurrent ({ current = null }) {
  if (current instanceof Element) return current
  throw new Error('Ref `current` is null')
}

const resizeObserver = new ResizeObserver(handleEntries)

const log = debug('@sequencemedia/graphviz-react')

export default function GraphvizReact ({
  graphRef: ref = createRef(),
  dot,
  className,
  options = DEFAULT_OPTIONS,
  onStart = DEFAULT_HANDLE_START,
  onRenderStart = DEFAULT_HANDLE_RENDER_START,
  onRenderEnd = DEFAULT_HANDLE_RENDER_END,
  onEnd = DEFAULT_HANDLE_END,
  onClick = DEFAULT_HANDLE_CLICK
}) {
  log('GraphvizReact')

  const [
    eventEmitter,
    setEventEmitter
  ] = useState(null)

  useEffect(() => {
    const {
      fit = false
    } = options

    if (fit) {
      if (hasCurrent(ref)) {
        const current = getCurrent(ref)

        resizeObserver.observe(current)

        return () => {
          resizeObserver.unobserve(current)
        }
      }
    }
  })

  useEffect(() => {
    if (hasCurrent(ref)) {
      const current = getCurrent(ref)

      const eventEmitter = (
        graphviz(current, {
          ...DEFAULT_OPTIONS,
          ...options
        })
          .renderDot(dot)
      )

      setEventEmitter(eventEmitter)
    }
  }, [dot, options])

  useEffect(() => {
    if (eventEmitter) {
      eventEmitter
        .on('start', onStart)
        .on('renderStart', onRenderStart)
        .on('renderEnd', onRenderEnd)
        .on('end', onEnd)
    }
  }, [eventEmitter, onStart, onRenderStart, onRenderEnd, onEnd])

  const handleClick = useCallback((event) => {
    if (hasEventTarget(event)) {
      const target = getEventTarget(event)

      if (hasCurrent(ref)) {
        const current = getCurrent(ref)

        if (current.contains(target)) onClick(event)
      }
    }
  }, [dot, options, onClick])

  return (
    <div
      className={classnames('graphviz', className)}
      onClick={handleClick}
      ref={ref}
    />
  )
}

GraphvizReact.propTypes = {
  graphRef: PropTypes.shape({
    current: PropTypes.shape()
  }),
  dot: PropTypes.string.isRequired,
  className: PropTypes.string,
  options: PropTypes.shape(),
  onClick: PropTypes.func,
  onStart: PropTypes.func,
  onRenderStart: PropTypes.func,
  onRenderEnd: PropTypes.func,
  onEnd: PropTypes.func
}
