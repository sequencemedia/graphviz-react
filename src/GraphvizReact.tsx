import React, {
  createRef,
  useState,
  useEffect,
  useCallback
} from 'react'
import classnames from 'classnames'
import type { GraphvizOptions } from 'd3-graphviz'
import { graphviz } from 'd3-graphviz'

export interface IGraphvizProps {
  /**
   *  A React `ref` for the container
   */
  graphRef?: React.RefObject<HTMLDivElement>
  /**
   * A string containing a graph representation using the Graphviz DOT language.
   * @see https://graphviz.org/doc/info/lang.html
   */
  dot: string
  /**
   * Options to pass to the Graphviz renderer.
   */
  options?: GraphvizOptions
  /**
   * The classname to attach to this component for styling purposes.
   */
  className?: string
  /**
   *  A handler for click events
   */
  onClick?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void
  /**
   *  A handler for `start` events
   */
  onStart?: () => void
  /**
   *  A handler for `renderStart` events
   */
  onRenderStart?: () => void
  /**
   *  A handler for `renderEnd` events
   */
  onRenderEnd?: () => void
  /**
   *  A handler for `end` events
   */
  onEnd?: () => void
}

const DEFAULT_OPTIONS: GraphvizOptions = {
  useWorker: false
}

export default function GraphvizReact ({
  graphRef: ref = createRef<HTMLDivElement>(),
  dot,
  className,
  options = {},
  onStart = (): void => {},
  onRenderStart = (): void => {},
  onRenderEnd = (): void => {},
  onEnd = (): void => {},
  onClick: handleClick = (): void => {}
}: IGraphvizProps): JSX.Element {
  const [eventEmitter, setEventEmitter] = useState<any>(null)

  useEffect(() => {
    const {
      current = null
    } = ref

    if (current !== null) {
      const eventEmitter: any = (
        graphviz(current, {
          ...DEFAULT_OPTIONS,
          ...options
        })
          .renderDot(dot)
      )

      setEventEmitter(eventEmitter)
    }
  }, [dot, options])

  useEffect(() => {
    if (eventEmitter !== null) {
      eventEmitter
        .on('start', onStart)
        .on('renderStart', onRenderStart)
        .on('renderEnd', onRenderEnd)
        .on('end', onEnd)
    }
  }, [eventEmitter, onStart, onRenderStart, onRenderEnd, onEnd])

  const onClick = useCallback((event: React.MouseEvent<HTMLDivElement, MouseEvent>): void => {
    const {
      current = null
    } = ref

    if (current !== null) {
      const {
        target = null
      } = event

      if (target instanceof Element) {
        if (current.contains(target)) {
          handleClick(event)
        }
      }
    }
  }, [dot, options, handleClick])

  return (
    <div
      className={classnames('graphviz', className)}
      onClick={onClick}
      ref={ref}
    />
  )
}
