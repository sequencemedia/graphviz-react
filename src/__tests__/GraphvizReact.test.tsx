import React from 'react'
import * as renderer from 'react-test-renderer'
import GraphvizReact from '../GraphvizReact.tsx'

describe('<GraphvizReact />', () => {
  it('matches the snapshot', () => {
    const graphviz = renderer.create(
      <GraphvizReact
        dot='graph { a -- b }'
      />
    )

    expect(graphviz.toJSON())
      .toMatchSnapshot()
  })
})
