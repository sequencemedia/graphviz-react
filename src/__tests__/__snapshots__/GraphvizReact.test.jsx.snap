// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GraphvizReact /> matches the snapshot 1`] = `
<div
  className="graphviz"
  onClick={[Function]}
/>
`;
